src/home/homework-blob.c

```diff
@@ -0,0 +1,248 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "copy.h"
+#include "fileio.h"
+#include "fd-util.h"
+#include "format-util.h"
+#include "fs-util.h"
+#include "home-util.h"
+#include "homework-blob.h"
+#include "homework.h"
+#include "install-file.h"
+#include "macro.h"
+#include "path-util.h"
+#include "recurse-dir.h"
+#include "rm-rf.h"
+#include "sha256.h"
+#include "string-util.h"
+#include "tmpfile-util.h"
+#include "umask-util.h"
+#include "utf8.h"
+
+static int copy_one_blob(
+                int src_fd,
+                int dest_dfd,
+                const char *name,
+                uint64_t *total_size,
+                uid_t uid,
+                Hashmap *manifest) {
+        _cleanup_(unlink_and_freep) char *dest_tmpname = NULL;
+        _cleanup_close_ int dest = -EBADF;
+        uint8_t hash[SHA256_DIGEST_SIZE], *known_hash;
+        off_t initial, size;
+        struct stat st;
+        int r;
+
+        assert(src_fd >= 0);
+        assert(dest_dfd >= 0);
+        assert(name);
+        assert(total_size);
+        assert(uid_is_valid(uid));
+        assert(manifest);
+
+        known_hash = hashmap_get(manifest, name);
+        if (!known_hash) {
+                log_warning("Blob %s is missing from manifest. Skipping.", name);
+                return 0;
+        }
+
+        if (!suitable_blob_filename(name)) {
+                log_warning("Blob %s has invalid filename. Skipping.", name);
+                return 0;
+        }
```

Comment:
frankly, i'd switch these checks, i.e. conceptually it makes sense to validate `name` first, and only then use it as key into the hashmap. i mean, this is technically correct both ways, but just conceptually it feels a bit weird using a string before it is validated.